### ISSUE TITLE
Datetimepicker not usable by keyboard

### ISSUE THREAD

**Issue by DasWiesel23** [Icinga/icingaweb2#4442]
Hello Team

I have an "issue" with the flatpickr calendar, which has been introduced in the latest icingaweb2 update.
Previously it was possible to click into the Start/End Time fields and adjust values manually in Chrome.
In Firefox you were able to type in a complete timestamp.

Even though I appreciate a browser-wide solution, I would also like to have the choice to adjust values by using the keyboard instead of clicking it together in flatpickr.

My question is, is there a quick fix on how to disable flatpickr user-wide or globally? I am using the opportunity to enter a complete timestamp like it is possible in Firefox very often.

Screenshot Firefox (old working):
![old_working_input_fields_firefox](https://user-images.githubusercontent.com/27570502/125505839-a776d462-25af-45de-af70-cbf85fcea368.png)

Screenshot Chromium (old working):
![old_working_input_fields_chromium](https://user-images.githubusercontent.com/27570502/125505842-cf85785c-691c-4726-a3b8-874538cc98ec.png)

Screenshot Chromium (not working):

![not_working_input_field](https://user-images.githubusercontent.com/27570502/125506186-f46c7711-f769-4da6-9a93-fddc32faad00.png)

Best,
Dennis

**Comment by lippserd** [comment]
I think we should rather check whether it’s possible to allow manual input. Anyway, thanks for the issue!

**Comment by nilmerg** [comment]
It's keyboard functionality isn't working correctly. Pasting a full date-time string into the field is a bonus feature but also possible.

**Comment by nilmerg** [comment]
>  is there a quick fix on how to disable flatpickr user-wide or globally?

Yes, apply the following change to `/usr/share/icingaweb2/application/views/helpers/FormDateTime.php`:

```diff
diff --git a/application/views/helpers/FormDateTime.php b/application/views/helpers/FormDateTime.php
index de5eb4b67..3b5ea7de0 100644
--- a/application/views/helpers/FormDateTime.php
+++ b/application/views/helpers/FormDateTime.php
@@ -49,7 +49,7 @@ class Zend_View_Helper_FormDateTime extends Zend_View_Helper_FormElement
         $type = $attribs['local'] === true ? 'datetime-local' : 'datetime';
         unset($attribs['local']);  // Unset local to not render it again in $this->_htmlAttribs($attribs)
         $html5 =  sprintf(
-            '<input type="%s" data-use-datetime-picker name="%s" id="%s" step="1" value="%s"%s%s%s',
+            '<input type="%s" name="%s" id="%s" step="1" value="%s"%s%s%s',
             $type,
             $this->view->escape($name),
             $this->view->escape($id),
```

**Comment by rafi0101** [comment]
I use icingaweb every day and would appreciate it if there would be again the possibility to enter the time via the keyboard, like in the previous version. Because with the old datetimepicker I was much faster in acknowledging and downtiming :)

**Comment by nilmerg** [comment]
I looked into this now but don't have good news unfortunately.

### General keyboard control

Is working for me with the current unstable of the flatpickr. Though, it's quite clunky I'd say.

To control the picker by keyboard, the input needs focus, then the key `down` must be pushed to enter the picker. Once entered, the arrow keys can be used to choose a date (day, month and year at the same time). Then pushing `Enter` gives focus to the hour input, which is a standard number input. (So arrow keys up/down work fine) The same applies to the minute input once pushing `Tab`. In case of 12 hour format, the am/pm selector is also reachable by pushing `Tab` again and changeable by the arrow keys up/down.

To close the picker, push `Esc` anytime while it or the input has focus. Or hit `Enter` while focusing the hour, minute and am/pm elements.

Improving this is difficult without patching flatpickr. There are [many issues regarding this](https://github.com/flatpickr/flatpickr/issues?q=is%3Aissue+is%3Aopen+keyboard+OR+accessible+) already.

### Enabling the input for user interaction

Possible. But useless, since there's no way to parse the **localized** format. The problem is that the `Intl.DateTimeFormat` we are using to format the date and time according to the user's locale, doesn't provide a way to parse a formatted string back again. It also doesn't provide a way to identify the format that's been used to format the value shown to the user.

Since the user interacts with the date and time formatted based on the locale, we'd have to parse this back to the server's format to both populate the picker and to send it to the server.

So, yes we can enable manual editing of the value, but as long as the format is localized there is no way to process the result. And disabling localization defeats the purpose of the picker. So, no manual interaction with the input for now.

### Pasting a ISO formatted string into the input

Yep, that we could parse and process. But without enabling manual edits in the input, the user cannot issue a paste event either. So how should we make pasting this intuitive if the usual thing (text input to paste text) isn't feasible? Enabling edits on the input you may ask... that's not really a good idea, since this opens the possibility to edit the content, which has no effect as said above. That's when users are irritated/frustrated and issues pop-up, demanding that the changes have an effect. And we're back to square one.

If someone shows me an intuitive and proven alternative to allow pasting a raw value in the picker, we may allow this.

### TL/DR

If manual interaction means that the picker is usable by keyboard, that's the only thing we can ensure. If it means to directly edit the form input, this is not possible without a complete rewrite of the entire implementation which only will add more complexity and hard-to-maintain solutions. (i.e. I don't want to take this path, and I hope keyboard interaction for the picker is sufficient for most users)

**Comment by rafi0101** [comment]
@nilmerg Is it an option to add the old datetimepicker in icingaweb and each user can decide in the user preferences which datetimepicker they want to use? 
Maybe the default is the new one, but everyone who wants to use the old one can change this in their settings?

**Comment by nilmerg** [comment]
We will only show the new datetime picker in Firefox. Other browsers have sufficient enough implementations themselves.

**Comment by chloesoe** [comment]
> We will only show the new datetime picker in Firefox. Other browsers have sufficient enough implementations themselves.

A way to disable the datepicker in FF would be cool. ATM if you use en_US in the browser, the date format is MM/D/YYYY and 12h clock format. 
I'd like to have an English interface but with normal Date format DD.MM.YYYY and 24h clock format in Firefox.
